### PR TITLE
fix: recipeUserId in sign in/up related APIs

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

- Link to issue1 here
- Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your
changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Update function `getValidFields` in `io/supertokens/config/CoreConfig.java` if new aliases were added for any core config (similar to the `access_token_signing_key_update_interval` config alias).
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: src/main/java/io/supertokens/webserver/api/passwordless/ConsumeCodeAPI.java [hunk 1]
```diff
@@ -108,6 +110,16 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
 
             result.addProperty("createdNewUser", consumeCodeResponse.createdNewUser);
             result.add("user", userJson);
+            if (getVersionFromRequest(req).greaterThanOrEqualTo(SemVer.v4_0)) {
+                for (LoginMethod loginMethod : consumeCodeResponse.user.loginMethods) {
+                    if (loginMethod.recipeId.equals(RECIPE_ID.PASSWORDLESS)
+                            && Objects.equals(loginMethod.email, consumeCodeResponse.email)
```
Comment (porcellus) [inline src/main/java/io/supertokens/webserver/api/passwordless/ConsumeCodeAPI.java:116]: Does this work if the pwless user has both email and phoneNumber associated with them (and they only log in using one)?

**Review by porcellus — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/webserver/api/passwordless/ConsumeCodeAPI.java (hunk 1 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/passwordless/ConsumeCodeAPI.java]: Implemented and added test for this case

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/webserver/api/emailpassword/SignUpAPI.java [hunk 2]
```diff
@@ -94,8 +94,10 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
             }
 
             result.add("user", userJson);
+            if (getVersionFromRequest(req).greaterThanOrEqualTo(SemVer.v4_0)) {
+                result.addProperty("recipeUserId", user.id);
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/emailpassword/SignUpAPI.java:98]: not required in signup api

File: src/main/java/io/supertokens/webserver/api/passwordless/ConsumeCodeAPI.java [hunk 3]
```diff
@@ -108,6 +110,16 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
 
             result.addProperty("createdNewUser", consumeCodeResponse.createdNewUser);
             result.add("user", userJson);
+            if (getVersionFromRequest(req).greaterThanOrEqualTo(SemVer.v4_0)) {
+                for (LoginMethod loginMethod : consumeCodeResponse.user.loginMethods) {
+                    if (loginMethod.recipeId.equals(RECIPE_ID.PASSWORDLESS)
+                            && (consumeCodeResponse.email == null || Objects.equals(loginMethod.email, consumeCodeResponse.email))
+                            && (consumeCodeResponse.phoneNumber == null || Objects.equals(loginMethod.phoneNumber, consumeCodeResponse.phoneNumber))) {
+                        result.addProperty("recipeUserId", loginMethod.recipeUserId);
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/passwordless/ConsumeCodeAPI.java]: done

File: src/main/java/io/supertokens/webserver/api/thirdparty/SignInUpAPI.java [hunk 4]
```diff
@@ -89,6 +91,16 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
                 }
 
                 result.add("user", userJson);
+                if (getVersionFromRequest(req).greaterThanOrEqualTo(SemVer.v4_0)) {
+                    for (LoginMethod loginMethod : response.user.loginMethods) {
+                        if (loginMethod.recipeId.equals(RECIPE_ID.THIRD_PARTY)
+                                && Objects.equals(loginMethod.thirdParty.id, thirdPartyId)
+                                && Objects.equals(loginMethod.thirdParty.userId, thirdPartyUserId)) {
+                            result.addProperty("recipeUserId", loginMethod.recipeUserId);
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/thirdparty/SignInUpAPI.java]: done

File: src/main/java/io/supertokens/webserver/api/thirdparty/SignInUpAPI.java [hunk 5]
```diff
@@ -141,6 +153,16 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
...
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/thirdparty/SignInUpAPI.java]: done

File: src/test/java/io/supertokens/test/accountlinking/api/TestRecipeUserIdInSignInUpAPIs.java [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+    @AfterClass
+    public static void afterTesting() {
+        Utils.afterTesting();
+    }
+
+    @Before
+    public void beforeEach() {
+        Utils.reset();
+    }
+
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/test/java/io/supertokens/test/accountlinking/api/TestRecipeUserIdInSignInUpAPIs.java:69]: done

File: src/main/java/io/supertokens/webserver/api/emailpassword/SignInAPI.java [hunk 7]
```diff
@@ -98,6 +99,15 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
                 userJson.remove("tenantIds");
             }
             result.add("user", userJson);
+            if (getVersionFromRequest(req).greaterThanOrEqualTo(SemVer.v4_0)) {
+                for (LoginMethod loginMethod : user.loginMethods) {
+                    if (loginMethod.recipeId.equals(RECIPE_ID.EMAIL_PASSWORD) && normalisedEmail.equals(loginMethod.email)) {
+                        result.addProperty("recipeUserId", loginMethod.recipeUserId);
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/emailpassword/SignInAPI.java]: done

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**